### PR TITLE
Deleted task projection

### PR DESCRIPTION
**Pull request by serhii-lekariev**
This PR extends the model of the to-do list by introducing `DeletedTaskProjection`

Before, when a task got deleted, it became inaccessible via the `MyListView`, and therefore ceased to be fetched forever.
The domain, however, according to its model, allows to restore deleted tasks, which was impossible, since client does not receive deleted tasks.

`DeletedTaskProjection` represents a single deleted task.

### CODE REVIEW

**Comment by serhii-lekariev** [comment]
@dmdashenkov, PTAL.

File: api-java/src/main/java/io/spine/examples/todolist/q/projection/DeletedTaskProjection.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    /**
+     * Creates a new instance.
+     *
+     * @param id
+     *         the ID for the new instance
+     * @throws IllegalArgumentException
+     *         if the ID is not of one of the supported types
+     */
+    protected DeletedTaskProjection(TaskId id) {
```
Comment (dmdashenkov) [inline api-java/src/main/java/io/spine/examples/todolist/q/projection/DeletedTaskProjection.java]: Entity constructors may be private.

File: api-java/src/main/java/io/spine/examples/todolist/q/projection/DeletedTaskProjection.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        super(id);
+    }
+
+    @Subscribe
+    public void on(TaskDeleted deleted, EventContext context) {
```
Comment (dmdashenkov) [inline api-java/src/main/java/io/spine/examples/todolist/q/projection/DeletedTaskProjection.java]: Subscriber methods should be package-private. This is a recent change. Please update all the projections within the project.

File: api-java/src/main/java/io/spine/examples/todolist/q/projection/DeletedTaskProjection.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        TaskId id = deleted.getTaskId();
+        TaskEnrichment enrichment = EnrichmentHelper.getEnrichment(TaskEnrichment.class, context);
```
Comment (dmdashenkov) [inline api-java/src/main/java/io/spine/examples/todolist/q/projection/DeletedTaskProjection.java]: As discussed, please remove `EnrichmentHelper`.

File: api-java/src/main/java/io/spine/examples/todolist/repository/DeletedTaskProjectionRepository.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * Repository for the {@link DeletedTaskProjection}.
+ */
+public class DeletedTaskProjectionRepository extends ProjectionRepository<TaskId,
+                                                                          DeletedTaskProjection,
+                                                                          DeletedTask> {
+
+    public DeletedTaskProjectionRepository() {
+        super();
+        rerouteEvents();
```
Comment (dmdashenkov) [inline api-java/src/main/java/io/spine/examples/todolist/repository/DeletedTaskProjectionRepository.java]: Please configure routing in `onRegistered()`. Let's make the constructors more concise.

File: api-java/src/test/java/io/spine/examples/todolist/q/projection/DeletedTaskProjectionTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                .newBuilder()
+                .setTaskId(taskId)
+                .build();
+        boundedContext.receivesEvent(restored);
+    }
+
+    @Test
+    @DisplayName("receive `TaskDeleted` event and set a respective ID")
+    void receiveAndGetId() {
+        TaskId taskToDelete = newTaskId();
```
Comment (dmdashenkov) [inline api-java/src/test/java/io/spine/examples/todolist/q/projection/DeletedTaskProjectionTest.java]: Please store the ID of the projection under test as a field (updated in `setUp()`) and remove the parameter from the event factory methods.

**Comment by serhii-lekariev** [comment]
@dmdashenkov, PTAL again,

File: api-java/src/main/java/io/spine/examples/todolist/q/projection/MyListViewProjection.java [hunk 6]
```diff
@@ -76,12 +76,12 @@
      * @throws IllegalArgumentException
      *         if the ID is not of one of the supported types
      */
-    public MyListViewProjection(TaskListId id) {
+    MyListViewProjection(TaskListId id) {
```
Comment (dmdashenkov) [inline api-java/src/main/java/io/spine/examples/todolist/q/projection/MyListViewProjection.java]: Please make it `private` is possible.

File: api-java/src/main/java/io/spine/examples/todolist/q/projection/MyListViewProjection.java [hunk 7]
```diff
@@ -103,74 +103,77 @@ public void on(TaskDeleted event) {
...
 
     @Subscribe
-    public void on(TaskDraftFinalized event, EventContext context) {
+    void on(TaskDraftFinalized event, EventContext context) {
         TaskId taskId = event.getTaskId();
-        TaskEnrichment enrichment = getEnrichment(TaskEnrichment.class, context);
-        Task task = enrichment.getTask();
+        Task task = context.find(TaskEnrichment.class)
+                           .map(TaskEnrichment::getTask)
+                           .orElseThrow(() -> new IllegalStateException(
```
Comment (dmdashenkov) [inline api-java/src/main/java/io/spine/examples/todolist/q/projection/MyListViewProjection.java]: Use `Exceptions.newIllegalStateException` instead.

File: api-java/src/main/java/io/spine/examples/todolist/q/projection/DeletedTaskProjection.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+import io.spine.examples.todolist.Task;
+import io.spine.examples.todolist.TaskDescription;
+import io.spine.examples.todolist.TaskId;
+import io.spine.examples.todolist.c.enrichments.TaskEnrichment;
+import io.spine.examples.todolist.c.events.DeletedTaskRestored;
+import io.spine.examples.todolist.c.events.TaskDeleted;
+import io.spine.server.projection.Projection;
+
+/**
+ * A task that has been as a result of a {@code TaskDeleted} event.
```
Comment (dmdashenkov) [inline api-java/src/main/java/io/spine/examples/todolist/q/projection/DeletedTaskProjection.java]: A task is not a result of an event.

Please reword.

File: api-java/src/main/java/io/spine/examples/todolist/q/projection/DeletedTaskProjection.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+public class DeletedTaskProjection extends Projection<TaskId, DeletedTask, DeletedTaskVBuilder> {
+
+    /**
+     * Creates a new instance.
+     *
+     * @param id
+     *         the ID for the new instance
+     * @throws IllegalArgumentException
+     *         if the ID is not of one of the supported types
+     */
```
Comment (dmdashenkov) [inline api-java/src/main/java/io/spine/examples/todolist/q/projection/DeletedTaskProjection.java]: Javadoc for these constructors is not required. You should leave this one here, but not feel guilty if you miss one.

File: api-java/src/main/java/io/spine/examples/todolist/q/projection/DeletedTaskProjection.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    private DeletedTaskProjection(TaskId id) {
+        super(id);
+    }
+
+    @Subscribe
+    void on(TaskDeleted deleted, EventContext context) {
+        TaskId id = deleted.getTaskId();
+        DeletedTaskVBuilder thisBuilder = builder();
```
Comment (dmdashenkov) [inline api-java/src/main/java/io/spine/examples/todolist/q/projection/DeletedTaskProjection.java]: I'd use just `builder`, not `thisBuilder` for the name.

File: api-java/src/main/java/io/spine/examples/todolist/q/projection/DeletedTaskProjection.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+     *         the ID for the new instance
+     * @throws IllegalArgumentException
+     *         if the ID is not of one of the supported types
+     */
+    private DeletedTaskProjection(TaskId id) {
+        super(id);
+    }
+
+    @Subscribe
+    void on(TaskDeleted deleted, EventContext context) {
```
Comment (dmdashenkov) [inline api-java/src/main/java/io/spine/examples/todolist/q/projection/DeletedTaskProjection.java]: What if a task is restored and then deleted again?
The projection handling this case would be the same as the one handling the first-time deletion. Please make sure that the value of `deleted` for this projection is `false` after `TaskDeleted` event.

File: api-java/src/main/java/io/spine/examples/todolist/q/projection/LabelledTasksViewProjection.java [hunk 12]
```diff
@@ -163,17 +164,20 @@ public void on(LabelDetailsUpdated event) {
...
-        TaskItem taskView = viewFor(taskDetails, labelId, taskId);
-        LabelDetails labelDetails = enrichment.getLabelDetails();
-
         builder().setId(labelId);
-        addTaskItem(taskView);
-        updateLabelDetails(labelDetails);
+        Optional<DetailsEnrichment> details = context.find(DetailsEnrichment.class);
+        if(!details.isPresent()){
+            String msg = "Could not get details enrichment from context %s.";
+            throw new IllegalStateException(format(msg, context));
```
Comment (dmdashenkov) [inline api-java/src/main/java/io/spine/examples/todolist/q/projection/LabelledTasksViewProjection.java]: Please use `Exceptions`.

File: api-java/src/test/java/io/spine/examples/todolist/q/projection/DeletedTaskProjectionTest.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+import io.spine.testing.server.blackbox.SingleTenantBlackBoxContext;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import static io.spine.examples.todolist.testdata.TestEventEnricherFactory.eventEnricherInstance;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@DisplayName("Deleted task projection should")
+public class DeletedTaskProjectionTest extends ProjectionTest {
```
Comment (dmdashenkov) [inline api-java/src/test/java/io/spine/examples/todolist/q/projection/DeletedTaskProjectionTest.java]: Please add a test for a task being deleted, restored, and then deleted again.

Comment (dmdashenkov) [reply api-java/src/test/java/io/spine/examples/todolist/q/projection/DeletedTaskProjectionTest.java]: Also, the test case class should be package-local.

File: model/src/main/proto/todolist/q/projections.proto [hunk 14]
```diff
@@ -104,6 +104,17 @@ message TaskListView {
     repeated TaskItem items = 1;
 }
 
+// A task that has been deleted as a result of a `TaskDeleted` event.
```
Comment (dmdashenkov) [inline model/src/main/proto/todolist/q/projections.proto]: Same as for the description of the Projection class.

**Comment by serhii-lekariev** [comment]
@dmdashenkov, PTAL again.

File: api-java/src/main/java/io/spine/examples/todolist/q/projection/LabelledTasksViewProjection.java [hunk 15]
```diff
@@ -59,8 +60,8 @@
  */
 @SuppressWarnings("OverlyCoupledClass")
 public class LabelledTasksViewProjection extends Projection<LabelId,
-                                                            LabelledTasksView,
-                                                            LabelledTasksViewVBuilder> {
+        LabelledTasksView,
+        LabelledTasksViewVBuilder> {
```
Comment (dmdashenkov) [inline api-java/src/main/java/io/spine/examples/todolist/q/projection/LabelledTasksViewProjection.java]: Please restore the formatting.

**Comment by serhii-lekariev** [comment]
@armiol, PTAL.

**Pull request merged**